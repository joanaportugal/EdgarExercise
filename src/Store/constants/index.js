export const ADD_POKEMON = "ADD_POKEMFETCH_POKEMON_SUCCESSON_ACTION";
export const USER_NAVIGATION = "NAVIGATION_ACTION";
export const ADD_POKEMON_LIST = "ADD_POKEMON_LIST";
export const USER_TYPE = "USER_TYPE";
export const FETCH_POKEMON_SUCCESS = "FETCH_POKEMON_SUCCESS";
export const FETCH_POKEMON_START = "FETCH_POKEMON_START"; 
export const FETCH_POKEMON_FAILURE = "FETCH_POKEMON_FAILURE";