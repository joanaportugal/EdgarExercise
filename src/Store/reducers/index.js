import {
  ADD_POKEMON,
  USER_NAVIGATION,
  ADD_POKEMON_LIST,
  USER_TYPE
} from "../constants";

const initialState = {
  url: { current: null, previous: null },
  list: { startId: null, endId: null, type: "All" },
  pokemonList: {}
};

export function pokemonReducer(state = initialState, action) {
  switch (action.type) {
    case ADD_POKEMON:
      return {
        ...state,
        pokemonList: { ...state.pokemonList, [action.name]: action.pokemon }
      };
    case USER_NAVIGATION:
      return {
        ...state,
        url: {
          current: action.url,
          previous: state.url.current
        }
      };
    case ADD_POKEMON_LIST:
      return {
        ...state,
        list: {
          ...state.list,
          startId: action.startId,
          endId: action.endId
        },
        pokemonList: {
          ...state.pokemonList,
          ...action.pokeList
        }
      };
    case USER_TYPE:
      return {
        ...state,
        list: {
          ...state.list,
          type: action.pokeType
        }
      };
    default:
      return state;
  }
}