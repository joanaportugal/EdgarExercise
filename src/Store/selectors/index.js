import { createSelector } from "reselect";

export const getCurrentPage = (state) => state.url?.current;

export const getPokemonID = createSelector(
    getCurrentPage, 
    page => {
    const url = page;
    if (!url) return null;
    const urlSplitted = url.split("/");
    return urlSplitted[1] === "reduxexercise" ? urlSplitted[2] : null;
});

export const getPokemons = state => state.pokemonList;

export const getPokemonById = createSelector(
    getPokemons,
    getPokemonID,
    (pokemons, id) => {
        const pokemonKey = Object.keys(pokemons).find(pokemon =>
            pokemons[pokemon].id == id)
        return pokemonKey ? pokemons[pokemonKey] : null
    }

);

export const getStartId = state => state.list.startId;

export const getEndId = state => state.list.endId; 

export const getType =  state => state.list.type;

export const getAllPokemonsByType = createSelector(
    getType,
    getPokemons,
    (selectedType, pokemons) => {
        if(selectedType === 'All') return pokemons
        let aux = {}
        Object.keys(pokemons).forEach( pokemonName => {
        if(pokemons[pokemonName].types.find(type => 
            type.type.name.toUpperCase() === selectedType.toUpperCase()
    )){
            aux = {...aux, [pokemonName] : pokemons[pokemonName]}
        }
      })
        return aux;
    }
)

export const getAllPokemonsOnList = createSelector(
    getAllPokemonsByType,
    getStartId, 
    getEndId,
    (pokemonsList, startId, endId) => {
        let aux = {}
        Object.keys(pokemonsList).forEach((pokemonName) => {
            if( pokemonsList[pokemonName].id >= startId && pokemonsList[pokemonName].id <= endId){
                aux = {...aux, [pokemonName]: pokemonsList[pokemonName]}
            } 
        })
        return aux;
    }
)