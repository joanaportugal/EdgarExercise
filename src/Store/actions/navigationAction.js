import { USER_NAVIGATION } from "../constants";

const navigationAction = url => ({
  type: USER_NAVIGATION,
  url
});

export default navigationAction;
