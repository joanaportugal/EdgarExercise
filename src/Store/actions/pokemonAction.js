import { ADD_POKEMON } from "../constants";

const pokemonAction = (name, pokemon) => ({
    type: ADD_POKEMON,
    name,
    pokemon
});

export default pokemonAction;