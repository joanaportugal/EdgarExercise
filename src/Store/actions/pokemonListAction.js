import { ADD_POKEMON_LIST, USER_TYPE } from "../constants";

export const pokemonListAction = (pokeList, startId, endId) => ({
  type: ADD_POKEMON_LIST,
  pokeList,
  startId,
  endId
});

export const userType = pokeType => ({
  type: USER_TYPE,
  pokeType
  
})