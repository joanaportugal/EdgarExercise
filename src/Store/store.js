import { createStore } from "redux";
import { pokemonReducer } from "./reducers/index";

const store = createStore(
  pokemonReducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);

export const createStoreWithInitialState = (initialState) => createStore(pokemonReducer, initialState)

export default store;
