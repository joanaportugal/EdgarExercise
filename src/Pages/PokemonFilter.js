import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { pokemonListAction, userType } from "../Store/actions/pokemonListAction.js";
import {getAllPokemonsOnList} from "../Store/selectors";
import {Card} from "../Components/Card";

const PokemonFilter = ({startId, endId}) => {
        const dispatch = useDispatch()
        const pokemonList = useSelector(getAllPokemonsOnList)
        console.log(pokemonList)
        const typeList = [
            "All",
            "Normal",
            "Poison",
            "Electric",
            "Fire",
            "Water",
            "Rock",
            "Grass",
            "Fighting",
            "Psychic",
            "Ground",
            "Fairy"
        ];
    

        async function fetchPokemon() {
            let pokemonListAux = {}
            for (let id = startId; id < endId; id++) {
                await fetch(`https://pokeapi.co/api/v2/pokemon/${id}/`)
                    .then(res => res.json())
                    .then(pokemon => {
                        pokemonListAux = {...pokemonListAux, [pokemon.name]: pokemon }
                    });
            }
            dispatch(pokemonListAction(pokemonListAux, startId, endId))
        }

        function getType(type, pokemonName) {
            switch (type) {
              case "speed":
                return pokemonList[pokemonName]?.stats[0]?.base_stat;
              case "specialDefense":
                return pokemonList[pokemonName]?.stats[1]?.base_stat;
              case "specialAttack":
                return pokemonList[pokemonName]?.stats[2]?.base_stat;
              case "defense":
                return pokemonList[pokemonName]?.stats[3]?.base_stat;
              case "attack":
                return pokemonList[pokemonName]?.stats[4]?.base_stat;
              case "hp":
                return pokemonList[pokemonName]?.stats[5]?.base_stat;
              case "weight":
                return pokemonList[pokemonName]?.weight;
              case "height":
                return pokemonList[pokemonName]?.height;
              case "type":
                return pokemonList[pokemonName]?.types[
                  pokemonList[pokemonName]?.types.length - 1
                ]?.type?.name;
              default:
                return;
            }
        }

        function getAvailabity(type, pokemonName) {
            const ability = pokemonList[pokemonName]?.abilities?.find(
              ability =>
                (type == "Normal" && !ability.is_hidden) ||
                (type == "Hidden" && ability.is_hidden)
            );
            return ability?.ability?.name;
        }

        useEffect(() => {
            if (Object.keys(pokemonList).length !== 0) return;
            fetchPokemon(startId, endId);
        });
          
        

        
        return ( 
            <>
                <ul>
                {
                    typeList.map((type, index) => (
                        <li key={index} onClick={() => dispatch(userType(type))}>{type}</li>
                    ))
                }
                </ul>
                
                {Object.keys(pokemonList).length !== 0 && 
                    Object.keys(pokemonList).map((pokemon, index) => (
                    <Card 
                        callbackAbility={() => null}
                        key={index} 
                        info={{
                            hp : getType("hp", pokemon),
                            attack : getType("attack", pokemon),
                            specialDefense: getType("specialDefense", pokemon),
                            defense: getType("defense", pokemon),
                            specialAttack: getType("specialAttack", pokemon),
                            speed: getType("speed", pokemon),
                            height: getType("height", pokemon),
                            weight: getType("weight", pokemon),
                            type: getType("type", pokemon),
                            
                        }} 
                        ability= {{ normal:getAvailabity("Normal", pokemon), hidden: getAvailabity("Hidden", pokemon) }}
                        imageSrc={pokemonList[pokemon]?.sprites?.front_default} 
                        name={pokemon}
                    />
                ))}
            </>
        )
}

export default PokemonFilter;