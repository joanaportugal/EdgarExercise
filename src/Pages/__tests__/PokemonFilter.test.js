import PokemonFilter from "../PokemonFilter";
import { render, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import React from "react";
import { createStoreWithInitialState } from "../../Store/store";
import * as actions from "../../Store/actions/pokemonListAction";

jest.spyOn(actions, "userType");

const state = {
  pokemonList: {
    pikachu: {
      id: 15,
      types: [{ type: { name: "electric" } }],
      stats: [
        {
          base_stat: 90,
          effort: 2,
          stat: {
            name: "speed",
            url: "https://pokeapi.co/api/v2/stat/6/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-defense",
            url: "https://pokeapi.co/api/v2/stat/5/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-attack",
            url: "https://pokeapi.co/api/v2/stat/4/"
          }
        },
        {
          base_stat: 40,
          effort: 0,
          stat: {
            name: "defense",
            url: "https://pokeapi.co/api/v2/stat/3/"
          }
        },
        {
          base_stat: 55,
          effort: 0,
          stat: {
            name: "attack",
            url: "https://pokeapi.co/api/v2/stat/2/"
          }
        },
        {
          base_stat: 35,
          effort: 0,
          stat: {
            name: "hp",
            url: "https://pokeapi.co/api/v2/stat/1/"
          }
        }
      ]
    },
    charizard: {
      id: 60,
      types: [{ type: { name: "fire" } }],
      stats: [
        {
          base_stat: 90,
          effort: 2,
          stat: {
            name: "speed",
            url: "https://pokeapi.co/api/v2/stat/6/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-defense",
            url: "https://pokeapi.co/api/v2/stat/5/"
          }
        },
        {
          base_stat: 50,
          effort: 0,
          stat: {
            name: "special-attack",
            url: "https://pokeapi.co/api/v2/stat/4/"
          }
        },
        {
          base_stat: 40,
          effort: 0,
          stat: {
            name: "defense",
            url: "https://pokeapi.co/api/v2/stat/3/"
          }
        },
        {
          base_stat: 55,
          effort: 0,
          stat: {
            name: "attack",
            url: "https://pokeapi.co/api/v2/stat/2/"
          }
        },
        {
          base_stat: 35,
          effort: 0,
          stat: {
            name: "hp",
            url: "https://pokeapi.co/api/v2/stat/1/"
          }
        }
      ]
    }
  },
  list: {
    startId: 10,
    endId: 80,
    type: "All"
  }
};

const store = createStoreWithInitialState(state);

describe("PokemonFilter", () => {
  it("component renders 2 card", () => {
    const { getByTestId, getAllByTestId, rerender } = render(
      <Provider store={store}>
        <PokemonFilter startId={null} endId={null} />
      </Provider>
    );

    expect(getAllByTestId("card").length).toBe(2);

    fireEvent.click(getAllByTestId("option")[3]);

    rerender(
      <Provider store={store}>
        <PokemonFilter startId={null} endId={null} />
      </Provider>
    );

    expect(actions.userType).toHaveBeenNthCalledWith(1, "Electric");

    expect(getAllByTestId("card").length).toBe(1);
  });

  it("snapshot", () => {
    const { container } = render(
      <Provider store={store}>
        <PokemonFilter startId={10} endId={20} />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
