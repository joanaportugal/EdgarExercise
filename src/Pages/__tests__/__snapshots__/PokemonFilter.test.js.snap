// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonFilter snapshot 1`] = `
<div>
  <ul>
    <li
      data-testid="option"
    >
      All
    </li>
    <li
      data-testid="option"
    >
      Normal
    </li>
    <li
      data-testid="option"
    >
      Poison
    </li>
    <li
      data-testid="option"
    >
      Electric
    </li>
    <li
      data-testid="option"
    >
      Fire
    </li>
    <li
      data-testid="option"
    >
      Water
    </li>
    <li
      data-testid="option"
    >
      Rock
    </li>
    <li
      data-testid="option"
    >
      Grass
    </li>
    <li
      data-testid="option"
    >
      Fighting
    </li>
    <li
      data-testid="option"
    >
      Psychic
    </li>
    <li
      data-testid="option"
    >
      Ground
    </li>
    <li
      data-testid="option"
    >
      Fairy
    </li>
  </ul>
  <div
    class="pokemonFilterList"
  >
    <div
      class="card u-full-width electric"
      data-testid="card"
    >
      <div
        class="imageWrapper"
      >
        <img
          class="pokemon-image"
        />
      </div>
      <div
        class="textWrapper"
      >
        <div
          class="pokemon-type"
        >
          electric
        </div>
        <div
          class="pokemon-name"
        >
          pikachu
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            HP:
          </div>
           
          <div
            class="pokemon-value"
          >
            35
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Attack:
          </div>
           
          <div
            class="pokemon-value"
          >
            55
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Defense:
          </div>
           
          <div
            class="pokemon-value"
          >
            40
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Special Attack:
          </div>
           
          <div
            class="pokemon-value"
          >
            50
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Special Defense:
          </div>
           
          <div
            class="pokemon-value"
          >
            50
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Speed:
          </div>
           
          <div
            class="pokemon-value"
          >
            90 m/s
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Height:
          </div>
           
          <div
            class="pokemon-value"
          >
            undefined feets
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="pokemon-column"
          >
            Weight:
          </div>
           
          <div
            class="pokemon-value"
          >
            undefined pounds
          </div>
        </div>
        <div
          class="info"
        >
          <div
            class="column"
          >
            <div
              class="pokemon-column"
            >
              Ability:
            </div>
            <div
              class="pokemon-value"
            />
          </div>
          <div
            class="column"
          >
            <div
              class="pokemon-column"
            >
              Hidden Ability:
            </div>
            <div
              class="pokemon-value"
            >
              --
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
